goal_src/jak2/engine/gfx/sprite/sprite.gc

```diff
@@ -5,5 +5,1073 @@
+(define *sprite-array-3d* (new 'global 'sprite-array-3d 256 0))
+
+(defun sprite-set-3d-quaternion! ((sprite sprite-vec-data-3d) (quat quaternion))
+  "Set the quaternion for a 3d sprite.
+   NOTE: the w component of the quaternion is not stored and
+   the original w value is preserved."
+  (local-vars (v1-0 float) (v1-1 float))
+  (rlet ((vf0 :class vf)
+         (vf1 :class vf)
+         (vf2 :class vf)
+         )
+    (init-vf0-vector)
+    (cond
+      ((< (-> quat w) 0.0)
+       (.lvf vf1 (&-> sprite qx-qy-qz-sy quad))
+       (.lvf vf2 (&-> quat vec quad))
+       (.sub.vf vf1 vf0 vf2 :mask #b111)
+       (.svf (&-> sprite qx-qy-qz-sy quad) vf1)
+       (.mov v1-0 vf1)
+       )
+      (else
+        (.lvf vf1 (&-> sprite qx-qy-qz-sy quad))
+        (.lvf vf2 (&-> quat vec quad))
+        (.add.vf vf1 vf0 vf2 :mask #b111)
+        (.svf (&-> sprite qx-qy-qz-sy quad) vf1)
+        (.mov v1-1 vf1)
+        )
+      )
+    quat
+    )
+  )
+
+(defun sprite-get-3d-quaternion! ((out quaternion) (sprite sprite-vec-data-3d))
+  "Get a quaternion. It's stored with only 3 components, so qw is
+   recalculated."
+  (let ((f0-0 (-> sprite qx-qy-qz-sy x))
+        (f1-0 (-> sprite qx-qy-qz-sy y))
+        (f2-0 (-> sprite qx-qy-qz-sy z))
+        )
+    (set! (-> out x) f0-0)
+    (set! (-> out y) f1-0)
+    (set! (-> out z) f2-0)
+    (set! (-> out w) (sqrtf (- (- (- 1.0 (* f2-0 f2-0)) (* f1-0 f1-0)) (* f0-0 f0-0))))
+    )
+  out
+  )
+
+(defun sprite-add-matrix-data ((dma-buff dma-buffer) (matrix-mode uint))
+  "Add matrix data to a dma-buffer.
+   Mode 0 = send (-> *math-camera* camera-temp)
+   Mode 1 = screen sprites, like the HUD"
+  (let ((count 900)) ;; maybe address??
+    (cond
+      ((zero? matrix-mode)
+       ;; upload to VU1 memory setup
+       (let* ((a1-1 dma-buff)
+              (pkt1 (the-as dma-packet (-> a1-1 base)))
+              )
+         (set! (-> pkt1 dma) (new 'static 'dma-tag :qwc #x5 :id (dma-tag-id cnt)))
+         (set! (-> pkt1 vif0) (new 'static 'vif-tag :imm #x404 :cmd (vif-cmd stcycl)))
+         (set! (-> pkt1 vif1) (new 'static 'vif-tag :num #x5 ;; matrix + qword
+                                                    :cmd (vif-cmd unpack-v4-32) 
+                                                    :imm (shr (shl count 54) 54)))
+         (set! (-> a1-1 base) (&+ (the-as pointer pkt1) 16))
+         )
+       ;; send the camera-temp matrix
+       (let* ((mtx (the-as matrix (-> dma-buff base)))
+              (t1-0 (-> *math-camera* camera-temp))
+              (a2-4 (-> t1-0 vector 0 quad))
+              (a3-4 (-> t1-0 vector 1 quad))
+              (t0-4 (-> t1-0 vector 2 quad))
+              (t1-1 (-> t1-0 trans quad))
+              )
+         (set! (-> mtx vector 0 quad) a2-4)
+         (set! (-> mtx vector 1 quad) a3-4)
+         (set! (-> mtx vector 2 quad) t0-4)
+         (set! (-> mtx trans quad) t1-1)
+         )
+       (&+! (-> dma-buff base) 64)
+       ;; also send this thing.
```

Comment:
`hvdf-off` = Horizontal, Vertical, Depth, Fog Offsets.
After the perspective transformation, this value is added the vector, giving the final horizontal, vertical, depth, and fog values.

Reply:
As a side note, they actually do fogging correctly, and don't fog based on depth, but instead actual distance from the camera.